### ISSUE TITLE
Eliminate `IsSqlBackend` constraint from various `withSqlConn` and `withSqlPool` functions

### ISSUE THREAD

**Issue by merijn** [yesodweb/persistent#783]
In #772 I created a new datatype to expose underlying SQLite internals more easily. However, due to the `IsSqlBackend` restriction on various functions in `Database.Persist.Sql` I find myself effectively reimplementing these functions to work with `RawSqlite` which is an instance of `PersistCore` and `HasPersistBackend` but *not* `IsSqlBackend` as that requires a `IsPersistBackend` instance which turns a `BaseBackend` (e.g. `SqlBackend`) into the relevant type. This can't be done, since the SQLite pointer can't be materialised from thin air.

At the same time, none of these functions seem to actually require an `IsSqlBackend` constraint as the typeclass isn't actually used. It's just there for the polymorphic return type, but then that can easily be propagated from the callers of `withSqlPool`. Removing these constraints would let these functions be reused for more types, avoiding the need to reimplement them.

**Comment by MaxGabriel** [comment]
Hm, that seems like it could be a big breaking change, wouldn't code like this (taken from the Yesod SQLite scaffolding) be broken, because the `pool` that `createSqlitePool` creates no longer has the `IsSqlBackend` class?

```haskell
    -- Create the database connection pool
    pool <- flip runLoggingT logFunc $ createSqlitePool
        (sqlDatabase $ appDatabaseConf appSettings)
        (sqlPoolSize $ appDatabaseConf appSettings)

    -- Perform database migration using our application's logging settings.
    runLoggingT (runSqlPool (runMigration migrateAll) pool) logFunc
```

Correct me if I'm wrong on this, just want to understand the full implications of this change

**Comment by merijn** [comment]
I'm only proposing the removal of the constraint on the generic functions inside `Database.Persist.Sql`, not the backend specific ones like `Database.Persist.Sqlite`. The constraint is *obviously* needed on, e.g. `withSqlitePool` and `createSqlitePool`. My observation is just that `Database.Persist.Sql.withSqlPool` (and friends) don't actually use the IsSqlBackend constraint, but due to having it can be reused to implement, for example, the `withRawSqlitePool` I proposed in #772 (because that doesn't fulfil the constraint). Backend specific variants like `withSqlitePool` should definitely keep the constraint, but if those have it, it becomes completely superfluous on `withSqlPool`.

**Comment by MaxGabriel** [comment]
@parsonsmatt You've made several changes to the Persistent typeclasses recently so I think you'd know better than me. Does this sound good to you?

**Comment by merijn** [comment]
I'll see where it can be safely (i.e. backwards compatibly) eliminated and create a PR somewhere this week.

**Comment by merijn** [comment]
Ah, actually, I just realised why these functions need those constraints. They, internally, call `close'` to close the SQL connection, so eliminating `IsSqlBackend` isn't feasible, but it could be generalised to either `HasPersistBackend backend, BaseBackend backend ~ SqlBackend`, `BackendCompatible SqlBackend backend`, or even `HasPersistBackend backend, BackendCompatible SqlBackend backend`. However, I'm not sure which of these would be the preferable approach, since the codebase seems mid-migration between two typeclass approaches?

If someone chimes in which way to go, I'll make sure to open a PR for that.

**Comment by parsonsmatt** [comment]
This is one of those times where I kinda wish Haskell had subtyping or row polymorphism or something :)

I made the `BackendCompatible` class to support the [`persistent-typed-db`](https://github.com/parsonsmatt/persistent-typed-db) library, which gives type-safe access to multiple SQL database schemas. The main problem I was running into was the conflation with `PersistEntityBackend record ~ BaseBackend backend`, and `BaseBackend` was tied with other things. Replacing that constraint with `PersistEntityBackend record ~ backend, BackendCompatible SqlBackend backend` allows you to say "this can be used as a `SqlBackend`" without the extraneous constraints.

**Comment by merijn** [comment]
@parsonsmatt So, do you have any opinion which of the 3 constraints I proposed would be best? Due to a paper deadline I'll have to postpone this until next week anyway, but probably shouldn't take me more than a day to actually go through the code and update everything IFF I know what I should actually update it to.

**Comment by MaxGabriel** [comment]
Pinging @parsonsmatt to respond to @merijn on this

**Comment by parsonsmatt** [comment]
I would prefer `BackendCompatible SqlBackend backend` to `(BaseBackend backend ~ SqlBackend)`, as it more directly communicates the intent.

**Comment by merijn** [comment]
ok, so I finally had some time to look into this and start writing some code. The number of changes is actually pretty small. The interplay between all the backends (or, at least SQLite and postgres) using `IsSqlBackend` and `BackendCompatible` is kinda annoying/tricky, though. Both are using `mkPersistBackend` to turn an `SqlBackend` into any `IsSqlBackend`.

The simplest solution for this is to just have `withSqlitePool` and friends simply have both a `BackendCompatible SqlBackend backend` and a `IsSqlBackend backend` constraint, but that seems rather ugly and redundant. But I also don't really see an easy way to get rid of the `IsSqlBackend` constraint without modifying the sqlite/postgres version of `withSqlPool` substantially.

**Comment by merijn** [comment]
Pinging @psibi @naushadh (because you two have the most recent commits pertaining to SQLite) and everyone in this thread again. I'd like to make some progress here, I have some code that's relying on my own fork now, but I'd like to consolidate that with upstream persistent-sqlite.

There's not a lot of work left to do here, but without some feedback from people I don't know how to finish this up...

**Comment by psibi** [comment]
@merijn Sorry that it took some time for me to get back to you. I read the whole thread above and I'm unable to understand the motivation of the problem.

> The simplest solution for this is to just have withSqlitePool and friends simply have both a BackendCompatible SqlBackend backend and a IsSqlBackend backend constraint, but that seems rather ugly and redundant. But I also don't really see an easy way to get rid of the IsSqlBackend constraint without modifying the sqlite/postgres version of withSqlPool substantially.

So are you proposing to add those constraints to the functions ? If yes, can you say whether adding those constraints will break existing code (like Yesod scaffolding site. Note that I'm not opposed to breaking change, I'm just trying to understand the motivation behind it). Also, by adding those constraints what type of code you are able to write which you are not able to achieve now. I would like to see a sample minimal code if possible. Thanks!

**Comment by merijn** [comment]
@psibi No worries. The motivation for this change is the work on https://github.com/yesodweb/persistent/pull/772 to expose the underlying SQLite pointer through persistent. Eliminating the IsSqlBackend constraint from these functions make it easier to generalise SQLite's `withSqlitePool` and friends to be more generic and support this new `RawSqlite` backend (which can't be made an instance of IsSqlBackend, as that requires the ability to turn an SqlBackend into RawSqlite, requires pulling the connection pointer from thin air).

**Comment by merijn** [comment]
This is obsoleted/solved by #886
